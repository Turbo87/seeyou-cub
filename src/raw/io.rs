use crate::error::Result;
use crate::types::ByteOrder;
use std::io::Read;

/// Read i16 with specified byte order
pub fn read_i16<R: Read>(reader: &mut R, order: ByteOrder) -> Result<i16> {
    let mut buf = [0u8; 2];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => i16::from_le_bytes(buf),
        ByteOrder::BE => i16::from_be_bytes(buf),
    })
}

/// Read u16 with specified byte order
pub fn read_u16<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u16> {
    let mut buf = [0u8; 2];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u16::from_le_bytes(buf),
        ByteOrder::BE => u16::from_be_bytes(buf),
    })
}

/// Read i32 with specified byte order
pub fn read_i32<R: Read>(reader: &mut R, order: ByteOrder) -> Result<i32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => i32::from_le_bytes(buf),
        ByteOrder::BE => i32::from_be_bytes(buf),
    })
}

/// Read u32 with specified byte order
pub fn read_u32<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u32::from_le_bytes(buf),
        ByteOrder::BE => u32::from_be_bytes(buf),
    })
}

/// Read u64 with specified byte order
pub fn read_u64<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u64> {
    let mut buf = [0u8; 8];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u64::from_le_bytes(buf),
        ByteOrder::BE => u64::from_be_bytes(buf),
    })
}

/// Read f32 (always little-endian per spec)
pub fn read_f32_le<R: Read>(reader: &mut R) -> Result<f32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(f32::from_le_bytes(buf))
}

/// Read u8
pub fn read_u8<R: Read>(reader: &mut R) -> Result<u8> {
    let mut buf = [0u8; 1];
    reader.read_exact(&mut buf)?;
    Ok(buf[0])
}

/// Read fixed-length byte array
pub fn read_bytes<R: Read>(reader: &mut R, len: usize) -> Result<Vec<u8>> {
    let mut buf = vec![0u8; len];
    reader.read_exact(&mut buf)?;
    Ok(buf)
}

/// Read string from bytes, attempting UTF-8 first, falling back to CP1252
pub fn read_string<R: Read>(reader: &mut R, len: usize) -> Result<String> {
    let bytes = read_bytes(reader, len)?;

    // Try UTF-8 first
    match String::from_utf8(bytes.clone()) {
        Ok(s) => Ok(s),
        Err(_) => {
            // Fallback to CP1252 (Windows-1252)
            let (decoded, _encoding, _had_errors) = encoding_rs::WINDOWS_1252.decode(&bytes);
            Ok(decoded.into_owned())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    #[test]
    fn test_read_i16_le() {
        let data = vec![0x34, 0x12];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i16(&mut cursor, ByteOrder::LE).unwrap(), 0x1234);
    }

    #[test]
    fn test_read_i16_be() {
        let data = vec![0x12, 0x34];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i16(&mut cursor, ByteOrder::BE).unwrap(), 0x1234);
    }

    #[test]
    fn test_read_i32_le() {
        let data = vec![0x78, 0x56, 0x34, 0x12];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i32(&mut cursor, ByteOrder::LE).unwrap(), 0x12345678);
    }

    #[test]
    fn test_read_f32_le() {
        let value = std::f32::consts::PI;
        let bytes = value.to_le_bytes();
        let mut cursor = Cursor::new(bytes);
        let result = read_f32_le(&mut cursor).unwrap();
        assert!((result - value).abs() < 0.0001);
    }

    #[test]
    fn test_read_string_utf8() {
        let data = b"Hello";
        let mut cursor = Cursor::new(data);
        assert_eq!(read_string(&mut cursor, 5).unwrap(), "Hello");
    }

    #[test]
    fn test_read_string_cp1252_fallback() {
        // CP1252 character (not valid UTF-8)
        let data = vec![0xE9]; // é in CP1252
        let mut cursor = Cursor::new(data);
        assert_eq!(read_string(&mut cursor, 1).unwrap(), "é");
    }
}
