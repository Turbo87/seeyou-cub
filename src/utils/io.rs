use crate::types::ByteOrder;
use std::io::{Read, Result, Write};

/// Read i16 with specified byte order
pub fn read_i16<R: Read>(reader: &mut R, order: ByteOrder) -> Result<i16> {
    let mut buf = [0u8; 2];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => i16::from_le_bytes(buf),
        ByteOrder::BE => i16::from_be_bytes(buf),
    })
}

/// Read u16 with specified byte order
pub fn read_u16<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u16> {
    let mut buf = [0u8; 2];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u16::from_le_bytes(buf),
        ByteOrder::BE => u16::from_be_bytes(buf),
    })
}

/// Read i32 with specified byte order
pub fn read_i32<R: Read>(reader: &mut R, order: ByteOrder) -> Result<i32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => i32::from_le_bytes(buf),
        ByteOrder::BE => i32::from_be_bytes(buf),
    })
}

/// Read u32 with specified byte order
pub fn read_u32<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u32::from_le_bytes(buf),
        ByteOrder::BE => u32::from_be_bytes(buf),
    })
}

/// Read u64 with specified byte order
pub fn read_u64<R: Read>(reader: &mut R, order: ByteOrder) -> Result<u64> {
    let mut buf = [0u8; 8];
    reader.read_exact(&mut buf)?;
    Ok(match order {
        ByteOrder::LE => u64::from_le_bytes(buf),
        ByteOrder::BE => u64::from_be_bytes(buf),
    })
}

/// Read f32 (always little-endian per spec)
pub fn read_f32_le<R: Read>(reader: &mut R) -> Result<f32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(f32::from_le_bytes(buf))
}

/// Read u8
pub fn read_u8<R: Read>(reader: &mut R) -> Result<u8> {
    let mut buf = [0u8; 1];
    reader.read_exact(&mut buf)?;
    Ok(buf[0])
}

/// Write u8
pub fn write_u8<W: Write>(writer: &mut W, value: u8) -> Result<()> {
    writer.write_all(&[value])
}

/// Write i16 with specified byte order
pub fn write_i16<W: Write>(writer: &mut W, value: i16, order: ByteOrder) -> Result<()> {
    let buf = match order {
        ByteOrder::LE => value.to_le_bytes(),
        ByteOrder::BE => value.to_be_bytes(),
    };
    writer.write_all(&buf)
}

/// Write u16 with specified byte order
pub fn write_u16<W: Write>(writer: &mut W, value: u16, order: ByteOrder) -> Result<()> {
    let buf = match order {
        ByteOrder::LE => value.to_le_bytes(),
        ByteOrder::BE => value.to_be_bytes(),
    };
    writer.write_all(&buf)
}

/// Write i32 with specified byte order
pub fn write_i32<W: Write>(writer: &mut W, value: i32, order: ByteOrder) -> Result<()> {
    let buf = match order {
        ByteOrder::LE => value.to_le_bytes(),
        ByteOrder::BE => value.to_be_bytes(),
    };
    writer.write_all(&buf)
}

/// Write u32 with specified byte order
pub fn write_u32<W: Write>(writer: &mut W, value: u32, order: ByteOrder) -> Result<()> {
    let buf = match order {
        ByteOrder::LE => value.to_le_bytes(),
        ByteOrder::BE => value.to_be_bytes(),
    };
    writer.write_all(&buf)
}

/// Write u64 with specified byte order
pub fn write_u64<W: Write>(writer: &mut W, value: u64, order: ByteOrder) -> Result<()> {
    let buf = match order {
        ByteOrder::LE => value.to_le_bytes(),
        ByteOrder::BE => value.to_be_bytes(),
    };
    writer.write_all(&buf)
}

/// Write f32 (always little-endian per spec)
pub fn write_f32_le<W: Write>(writer: &mut W, value: f32) -> Result<()> {
    writer.write_all(&value.to_le_bytes())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    #[test]
    fn test_read_i16_le() {
        let data = vec![0x34, 0x12];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i16(&mut cursor, ByteOrder::LE).unwrap(), 0x1234);
    }

    #[test]
    fn test_read_i16_be() {
        let data = vec![0x12, 0x34];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i16(&mut cursor, ByteOrder::BE).unwrap(), 0x1234);
    }

    #[test]
    fn test_read_i32_le() {
        let data = vec![0x78, 0x56, 0x34, 0x12];
        let mut cursor = Cursor::new(data);
        assert_eq!(read_i32(&mut cursor, ByteOrder::LE).unwrap(), 0x12345678);
    }

    #[test]
    fn test_read_f32_le() {
        let value = std::f32::consts::PI;
        let bytes = value.to_le_bytes();
        let mut cursor = Cursor::new(bytes);
        let result = read_f32_le(&mut cursor).unwrap();
        assert!((result - value).abs() < 0.0001);
    }

    #[test]
    fn test_write_u8() {
        let mut buf = Vec::new();
        write_u8(&mut buf, 0x42).unwrap();
        assert_eq!(buf, vec![0x42]);
    }

    #[test]
    fn test_write_i16_le() {
        let mut buf = Vec::new();
        write_i16(&mut buf, 0x1234, ByteOrder::LE).unwrap();
        assert_eq!(buf, vec![0x34, 0x12]);
    }

    #[test]
    fn test_write_i16_be() {
        let mut buf = Vec::new();
        write_i16(&mut buf, 0x1234, ByteOrder::BE).unwrap();
        assert_eq!(buf, vec![0x12, 0x34]);
    }

    #[test]
    fn test_write_u16_le() {
        let mut buf = Vec::new();
        write_u16(&mut buf, 0x1234, ByteOrder::LE).unwrap();
        assert_eq!(buf, vec![0x34, 0x12]);
    }

    #[test]
    fn test_write_u16_be() {
        let mut buf = Vec::new();
        write_u16(&mut buf, 0x1234, ByteOrder::BE).unwrap();
        assert_eq!(buf, vec![0x12, 0x34]);
    }

    #[test]
    fn test_write_i32_le() {
        let mut buf = Vec::new();
        write_i32(&mut buf, 0x12345678, ByteOrder::LE).unwrap();
        assert_eq!(buf, vec![0x78, 0x56, 0x34, 0x12]);
    }

    #[test]
    fn test_write_i32_be() {
        let mut buf = Vec::new();
        write_i32(&mut buf, 0x12345678, ByteOrder::BE).unwrap();
        assert_eq!(buf, vec![0x12, 0x34, 0x56, 0x78]);
    }

    #[test]
    fn test_write_u32_le() {
        let mut buf = Vec::new();
        write_u32(&mut buf, 0x12345678, ByteOrder::LE).unwrap();
        assert_eq!(buf, vec![0x78, 0x56, 0x34, 0x12]);
    }

    #[test]
    fn test_write_u32_be() {
        let mut buf = Vec::new();
        write_u32(&mut buf, 0x12345678, ByteOrder::BE).unwrap();
        assert_eq!(buf, vec![0x12, 0x34, 0x56, 0x78]);
    }

    #[test]
    fn test_write_u64_le() {
        let mut buf = Vec::new();
        write_u64(&mut buf, 0x123456789ABCDEF0, ByteOrder::LE).unwrap();
        assert_eq!(buf, vec![0xF0, 0xDE, 0xBC, 0x9A, 0x78, 0x56, 0x34, 0x12]);
    }

    #[test]
    fn test_write_u64_be() {
        let mut buf = Vec::new();
        write_u64(&mut buf, 0x123456789ABCDEF0, ByteOrder::BE).unwrap();
        assert_eq!(buf, vec![0x12, 0x34, 0x56, 0x78, 0x9A, 0xBC, 0xDE, 0xF0]);
    }

    #[test]
    fn test_write_f32_le() {
        let value = std::f32::consts::PI;
        let mut buf = Vec::new();
        write_f32_le(&mut buf, value).unwrap();
        assert_eq!(buf, value.to_le_bytes());
    }
}
