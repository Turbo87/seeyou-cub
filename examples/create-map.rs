#![expect(unused_imports)]

use seeyou_cub::{Airspace, CubReader, CubStyle};
use serde_json::json;
use std::env;
use std::fs;
use std::io;

/// Maps airspace style to hex color following aviation conventions
fn style_to_color(style: &CubStyle) -> &'static str {
    match style {
        CubStyle::ProhibitedArea => "#8B0000",           // Dark red
        CubStyle::RestrictedArea => "#DC143C",           // Red
        CubStyle::DangerArea => "#FF4500",               // Orange red
        CubStyle::ControlZone => "#4169E1",              // Blue
        CubStyle::TerminalControlArea => "#87CEEB",      // Light blue
        CubStyle::ControlArea => "#9370DB",              // Purple
        CubStyle::GliderSector => "#FFD700",             // Yellow
        CubStyle::TransponderMandatoryZone => "#FF8C00", // Dark orange
        CubStyle::Warning => "#FFA500",                  // Orange
        CubStyle::Alert => "#FF6347",                    // Tomato
        _ => "#808080",                                  // Gray (unknown/other)
    }
}

/// Converts an Airspace to a GeoJSON Feature
#[cfg_attr(not(test), allow(dead_code))]
fn airspace_to_geojson_feature(airspace: &Airspace) -> serde_json::Value {
    // Convert points: radians -> degrees, lat/lon -> [lon, lat]
    let mut coordinates: Vec<[f32; 2]> = airspace
        .points
        .iter()
        .map(|pt| [pt.lon.to_degrees(), pt.lat.to_degrees()])
        .collect();

    // Ensure polygon is closed (first point = last point)
    if let Some(first) = coordinates.first().copied()
        && coordinates.last() != Some(&first)
    {
        coordinates.push(first);
    }

    json!({
        "type": "Feature",
        "geometry": {
            "type": "Polygon",
            "coordinates": [coordinates]
        },
        "properties": {
            "name": airspace.name,
            "class": format!("{:?}", airspace.class),
            "style": format!("{:?}", airspace.style),
            "min_alt": airspace.min_alt,
            "max_alt": airspace.max_alt,
            "min_alt_style": format!("{:?}", airspace.min_alt_style),
            "max_alt_style": format!("{:?}", airspace.max_alt_style),
            "color": style_to_color(&airspace.style)
        }
    })
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Parse command-line arguments
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: create-map <path-to-cub-file>");
        std::process::exit(1);
    }

    let _cub_path = &args[1];

    // TODO: Implementation goes here

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use seeyou_cub::{AltStyle, CubClass, DaysActive, Point};

    #[test]
    fn test_style_to_color() {
        assert_eq!(style_to_color(&CubStyle::ProhibitedArea), "#8B0000");
        assert_eq!(style_to_color(&CubStyle::RestrictedArea), "#DC143C");
        assert_eq!(style_to_color(&CubStyle::DangerArea), "#FF4500");
        assert_eq!(style_to_color(&CubStyle::ControlZone), "#4169E1");
        assert_eq!(style_to_color(&CubStyle::GliderSector), "#FFD700");
    }

    #[test]
    fn test_airspace_to_geojson_feature() {
        let airspace = Airspace {
            name: "Test Zone".to_string(),
            points: vec![
                Point { lat: 0.5, lon: 0.4 }, // ~28.6° lat, ~22.9° lon
                Point { lat: 0.6, lon: 0.5 },
                Point { lat: 0.5, lon: 0.6 },
            ],
            style: CubStyle::ControlZone,
            class: CubClass::ClassD,
            min_alt: 0,
            max_alt: 5000,
            min_alt_style: AltStyle::MeanSeaLevel,
            max_alt_style: AltStyle::MeanSeaLevel,
            days_active: DaysActive::all(),
            ..Default::default()
        };

        let feature = airspace_to_geojson_feature(&airspace);
        assert_debug_snapshot!(feature);
    }
}
