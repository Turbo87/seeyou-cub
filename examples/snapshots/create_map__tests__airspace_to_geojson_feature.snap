---
source: examples/create-map.rs
expression: feature
---
Object {
    "geometry": Object {
        "coordinates": Array [
            Array [
                Array [
                    Number(22.918312072753906),
                    Number(28.647890090942383),
                ],
                Array [
                    Number(28.647890090942383),
                    Number(34.37746810913086),
                ],
                Array [
                    Number(34.37746810913086),
                    Number(28.647890090942383),
                ],
                Array [
                    Number(22.918312072753906),
                    Number(28.647890090942383),
                ],
            ],
        ],
        "type": String("Polygon"),
    },
    "properties": Object {
        "class": String("ClassD"),
        "color": String("#4169E1"),
        "max_alt": Number(5000),
        "max_alt_style": String("MeanSeaLevel"),
        "min_alt": Number(0),
        "min_alt_style": String("MeanSeaLevel"),
        "name": String("Test Zone"),
        "style": String("ControlZone"),
    },
    "type": String("Feature"),
}
